### PR TITLE
vendor: github.com/moby/buildkit v0.17.0-rc1

### PR DESCRIPTION
**Pull request by thaJeztah**
- supersedes / closes https://github.com/moby/moby/pull/48623

full diff:

- https://github.com/moby/buildkit/compare/v0.16.0...v0.17.0-rc1
- https://github.com/tonistiigi/fsutil/compare/91a3fc46842c...3f140a1299b0
- https://github.com/golang/exp/compare/9b4947da3948...701f63a606c0

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**

```markdown changelog
Update BuildKit to [v0.17.0-rc1](https://github.com/moby/buildkit/releases/tag/v0.17.0-rc1)
```

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
Some changes needed here, related to https://github.com/moby/buildkit/pull/5342

https://github.com/moby/moby/blob/c09e5265db33f3141713f3094a10bc80025ee739/builder/builder-next/builder.go#L165-L167

- `Size_` field was renamed to `Size` (no trailing `_`)
- `CreatedAt` field changed from a `time.Time` to a `*timestamp.Timestamp` (so both changed to a pointer, and a different type)
- `CreatedAt` field changed from a `*time.Time` to `*timestamp.Timestamp` (different type)

**Comment by thaJeztah** [comment]
> Sigh...
> 
> ```
> 33.49 # github.com/docker/docker/builder/builder-next/worker
> 33.49 builder/builder-next/worker/gc.go:35:4: unknown field KeepBytes in struct literal of type struct{All bool "json:\"all\""; Filter []string "json:\"filter\""; KeepDuration "time".Duration "json:\"keepDuration\""; ReservedSpace int64 "json:\"reservedSpace\""; MaxUsedSpace int64 "json:\"maxUsedSpace\""; MinFreeSpace int64 "json:\"minFreeSpace\""}
> 33.49 builder/builder-next/worker/gc.go:40:4: unknown field KeepBytes in struct literal of type struct{All bool "json:\"all\""; Filter []string "json:\"filter\""; KeepDuration "time".Duration "json:\"keepDuration\""; ReservedSpace int64 "json:\"reservedSpace\""; MaxUsedSpace int64 "json:\"maxUsedSpace\""; MinFreeSpace int64 "json:\"minFreeSpace\""}
> 33.49 builder/builder-next/worker/gc.go:44:4: unknown field KeepBytes in struct literal of type struct{All bool "json:\"all\""; Filter []string "json:\"filter\""; KeepDuration "time".Duration "json:\"keepDuration\""; ReservedSpace int64 "json:\"reservedSpace\""; MaxUsedSpace int64 "json:\"maxUsedSpace\""; MinFreeSpace int64 "json:\"minFreeSpace\""}
> 33.49 builder/builder-next/worker/gc.go:49:4: unknown field KeepBytes in struct literal of type struct{All bool "json:\"all\""; Filter []string "json:\"filter\""; KeepDuration "time".Duration "json:\"keepDuration\""; ReservedSpace int64 "json:\"reservedSpace\""; MaxUsedSpace int64 "json:\"maxUsedSpace\""; MinFreeSpace int64 "json:\"minFreeSpace\""}
> ```

Those changes relate to;

- https://github.com/moby/buildkit/pull/5359

Changes;

- The  `KeepBytes` field was renamed to `ReservedSpace`
- New field `MaxUsedSpace` added
- New field `MinFreeSpace` added

Old Type:

https://github.com/moby/moby/blob/c09e5265db33f3141713f3094a10bc80025ee739/vendor/github.com/moby/buildkit/client/prune.go#L61-L66


New Type;

https://github.com/moby/buildkit/blob/2534310fd4c59018ae3874e8d0fad493086e2575/client/prune.go#L69-L77


```go
type PruneInfo struct {
	All          bool          `json:"all"`
	Filter       []string      `json:"filter"`
	KeepDuration time.Duration `json:"keepDuration"`

	ReservedSpace int64 `json:"reservedSpace"`
	MaxUsedSpace  int64 `json:"maxUsedSpace"`
	MinFreeSpace  int64 `json:"minFreeSpace"`
}
```


The `PruneRequest` type in the API also changed;

- The  `KeepBytes` field was renamed to `ReservedSpace`
- New field `MaxUsedSpace` added
- New field `MinFreeSpace` added


Old: https://github.com/moby/moby/blob/c09e5265db33f3141713f3094a10bc80025ee739/vendor/github.com/moby/buildkit/api/services/control/control.pb.go#L68-L76

New: https://github.com/moby/buildkit/blob/2534310fd4c59018ae3874e8d0fad493086e2575/api/services/control/control.pb.go

```go
type PruneRequest struct {
	state         protoimpl.MessageState
	sizeCache     protoimpl.SizeCache
	unknownFields protoimpl.UnknownFields

	Filter        []string `protobuf:"bytes,1,rep,name=filter,proto3" json:"filter,omitempty"`
	All           bool     `protobuf:"varint,2,opt,name=all,proto3" json:"all,omitempty"`
	KeepDuration  int64    `protobuf:"varint,3,opt,name=keepDuration,proto3" json:"keepDuration,omitempty"`
	ReservedSpace int64    `protobuf:"varint,4,opt,name=reservedSpace,proto3" json:"reservedSpace,omitempty"`
	MaxUsedSpace  int64    `protobuf:"varint,5,opt,name=maxUsedSpace,proto3" json:"maxUsedSpace,omitempty"`
	MinFreeSpace  int64    `protobuf:"varint,6,opt,name=minFreeSpace,proto3" json:"minFreeSpace,omitempty"`
}
```

**Comment by thaJeztah** [comment]
Ah; more API changes; `SolveRequest.Cache` changed from a non-pointer

https://github.com/moby/moby/blob/c09e5265db33f3141713f3094a10bc80025ee739/vendor/github.com/moby/buildkit/api/services/control/control.pb.go#L380

To a pointer; https://github.com/moby/buildkit/blob/2534310fd4c59018ae3874e8d0fad493086e2575/api/services/control/control.pb.go#L410

```go
	Cache                   *CacheOptions             `protobuf:"bytes,8,opt,name=Cache,proto3" json:"Cache,omitempty"`
```

File: builder/builder-next/builder.go [hunk 1]
```diff
@@ -437,7 +445,7 @@ func (b *Builder) Build(ctx context.Context, opt backend.BuildConfig) (*builder.
 
 	eg.Go(func() error {
 		for sr := range ch {
-			dt, err := sr.Marshal()
+			dt, err := proto.Marshal(sr) // FIXME(thaJeztah): is this the correct replacement?
```
Comment (thaJeztah) [inline builder/builder-next/builder.go:448]: Don't know if this is the right replacement; if someone knows how to verify this?

File: builder/builder-next/builder.go [hunk 2]
```diff
@@ -624,6 +632,7 @@ func toBuildkitUlimits(inp []*container.Ulimit) (string, error) {
 	return strings.Join(ulimits, ","), nil
 }
 
+// FIXME(thaJeztah): wire-up new fields; see https://github.com/moby/moby/issues/48639
```
Comment (thaJeztah) [inline builder/builder-next/builder.go:635]: I decided to keep the daemon-config as it was, so only made the changes internal for now, but added a bunch of TODO's / FIXME's, and a ticket to look at what's needed to implement the new options and deprecate the old ones;

- https://github.com/moby/moby/issues/48639

File: builder/builder-next/builder.go [hunk 3]
```diff
@@ -162,16 +163,23 @@ func (b *Builder) DiskUsage(ctx context.Context) ([]*types.BuildCache, error) {
...
-			LastUsedAt:  r.LastUsedAt,
-			UsageCount:  int(r.UsageCount),
+			Size:        r.Size,
+			CreatedAt: func() time.Time {
+				if r.CreatedAt != nil {
+					return r.CreatedAt.AsTime()
+				}
+				return time.Time{}
+			}(),
+			LastUsedAt: func() *time.Time { t := r.LastUsedAt.AsTime(); return &t }(),
```
Comment (thaJeztah) [inline builder/builder-next/builder.go:173]: This is a bit hairy; also not sure why one was a pointer and the other one wasn't.

Comment (thaJeztah) [reply builder/builder-next/builder.go:173]: Oh! I need to do a nil-check on `LastUsedAt` as well; not sure if it could ever be `nil`.

**Comment by thaJeztah** [comment]
At least CI is green; let me update to add that nil-check, and push again

File: builder/builder-next/builder.go (hunk 1 above)
Comment (jsternberg) [reply builder/builder-next/builder.go:454]: This is correct.

File: builder/builder-next/builder.go (hunk 3 above)
Comment (jsternberg) [reply builder/builder-next/builder.go]: Honestly not sure why some of these were pointers and some weren't from the original when removing gogo proto. I just kind of went with it.

**Review by jsternberg — APPROVED** [review]
I can follow up on wiring the gc changes and introducing vtproto for the grpc serialization. This looks good.

File: builder/builder-next/builder.go (hunk 1 above)
Comment (thaJeztah) [reply builder/builder-next/builder.go:454]: Thanks! Let me remove this FIXME 👍

**Comment by thaJeztah** [comment]
Rebased, and removed the "FIXME" comment from https://github.com/moby/moby/pull/48634#discussion_r1797194091


To others; this was green before I pushed, and probably should be ready to merge once CI finishes

**Review by rumpl — APPROVED** [review]

**Review by jsternberg — APPROVED** [review]

**Comment by thaJeztah** [comment]
I need to tweak those timeouts on GHA. For some reason, I had it 3 times today where `bin-image` or `ci/cross` just hung; nothing happening, but also no way to cancel the job. With the timeouts I added, at least it's no longer than 2 hours, but given that those jobs normally take less than 5 minutes, I should tweak those.

**Comment by thaJeztah** [comment]
FWIW; this was the failing one; https://github.com/moby/moby/actions/runs/11296042961/attempts/1 -> https://github.com/moby/moby/actions/runs/11296042961/job/31420083939

And the logs; 
[logs_29517508817.zip](https://github.com/user-attachments/files/17346774/logs_29517508817.zip)

**Pull request merged**